# Binary-Data-Classification-with-ANN

Installation and set up 

In [129]:
# install tensoflow
!pip install tensorflow 

In [130]:
# import and check the version of the tensorflow 
import tensorflow as tf
print(tf.__version__)

2.0.0


In [131]:
# import other libararies 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [132]:
dataset=pd.read_csv('Churn_Modelling.csv')
dataset.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [133]:
# selecting the independent and dependent variables 
# there are 2 axiss, "0" for to denotes rows, "1" for denote the columns
x = dataset.drop(labels=['RowNumber', 'CustomerId', 'Surname','Exited'], axis=1)
y = dataset['Exited']

In [134]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [135]:
x.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [136]:
# Encoding the catergorical values 
from sklearn.preprocessing import LabelEncoder

In [137]:
# create an object
lblEncoder1 = LabelEncoder()
# first fit to the created object then transform it back to the original column (replace)
x['Geography'] = lblEncoder1.fit_transform(x['Geography'])
x.head()

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619          0  Female   42       2       0.00              1   
1          608          2  Female   41       1   83807.86              1   
2          502          0  Female   42       8  159660.80              3   
3          699          0  Female   39       1       0.00              2   
4          850          2  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [138]:
lblEncoder2 = LabelEncoder()
x['Gender'] = lblEncoder2.fit_transform(x["Gender"])
x.head()

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619          0       0   42       2       0.00              1   
1          608          2       0   41       1   83807.86              1   
2          502          0       0   42       8  159660.80              3   
3          699          0       0   39       1       0.00              2   
4          850          2       0   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

Even it's encoded we can't compare(france>spain) Geography and Gender columns. So we have to avoid this trap 

In [139]:
# Avoiding the varibles trap 
x= pd.get_dummies(x, drop_first=True, columns=['Geography'])
x.head(8)

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619       0   42       2       0.00              1          1   
1          608       0   41       1   83807.86              1          0   
2          502       0   42       8  159660.80              3          1   
3          699       0   39       1       0.00              2          0   
4          850       0   43       2  125510.82              1          1   
5          645       1   44       8  113755.78              2          1   
6          822       1   50       7       0.00              2          1   
7          376       0   29       4  115046.74              4          1   

   IsActiveMember  EstimatedSalary  Geography_1  Geography_2  
0               1        101348.88            0            0  
1               1        112542.58            0            1  
2               0        113931.57            0            0  
3               0         93826.63            0            0  
4               1         79084.10            0            1  
5               0        149756.71            0            1  
6               1         10062.80            0            0  
7               0        119346.88            1            0

In [140]:
# split the data to train and test 
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=0)

In [141]:
# Feature scaling 
# to avoid the effect of the large scaling 
# ex:
x.head(1)
# values are diffrent from  619 - > 0 ->42 .. spread over the large scale, to reduce the impact it is going to 

CreditScore  Gender  Age  Tenure  Balance  NumOfProducts  HasCrCard  \
0          619       0   42       2      0.0              1          1   

   IsActiveMember  EstimatedSalary  Geography_1  Geography_2  
0               1        101348.88            0            0

In [142]:
# Feature Scaling (bring value scale of 1 to 1)
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
# no need to fit data because already fitted to the sc object 
x_test = sc.transform(x_test)

In [143]:
x_train

array([[ 0.16958176, -1.09168714, -0.46460796, ...,  1.10643166,
        -0.5698444 ,  1.74309049],
       [-2.30455945,  0.91601335,  0.30102557, ..., -0.74866447,
         1.75486502, -0.57369368],
       [-1.19119591, -1.09168714, -0.94312892, ...,  1.48533467,
        -0.5698444 , -0.57369368],
       ...,
       [ 0.9015152 ,  0.91601335, -0.36890377, ...,  1.41231994,
        -0.5698444 , -0.57369368],
       [-0.62420521, -1.09168714, -0.08179119, ...,  0.84432121,
        -0.5698444 ,  1.74309049],
       [-0.28401079, -1.09168714,  0.87525072, ...,  0.32472465,
         1.75486502, -0.57369368]])

# Build the model 

In [144]:
# create the model 
# To build the model it uses Sequential calss beacue it can use to create fully connected neural network 
model = tf.keras.models.Sequential()

In [145]:
# Adding input and hidden layer 
# unit = 6 
# activation function = ReLu
# input dimention = 11 

# the reason to take 6 units is, there are 11 inputs and 1 outout, all together is 12 and taken the avarage by 12/2
# the above unit selection is not a rule, we can use any number, above one is only a trick 

model.add(tf.keras.layers.Dense(units =6, activation= "relu", input_dim=11))

In [146]:
x.shape

(10000, 11)

In [147]:
# Adding hiiden layers
model.add(tf.keras.layers.Dense(units=6, activation="relu"))

In [148]:
# out put layer
# one unit becasue we need only one output 
# use sigmoid, it is good for binary classifiactions
model.add(tf.keras.layers.Dense(units =1, activation="sigmoid"))

In [149]:
# compile the model 
model.compile(optimizer="adam", loss="binary_crossentropy", metrics=["accuracy"])

In [150]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_8 (Dense)              (None, 6)                 72        
_________________________________________________________________
dense_9 (Dense)              (None, 6)                 42        
_________________________________________________________________
dense_10 (Dense)             (None, 1)                 7         
Total params: 121
Trainable params: 121
Non-trainable params: 0
_________________________________________________________________


# Train the model 

In [153]:
# x_train has 11 colums and it's a numpy array, y_train has only a 1 column, so it's requied to convert convert into numpy array
model.fit(x_train, y_train.to_numpy(), batch_size=10, epochs=20)

Train on 8000 samples
Epoch 1/20
8000/8000 [==============================] - 2s 223us/sample - loss: 0.4911 - accuracy: 0.7960
Epoch 2/20
8000/8000 [==============================] - 1s 120us/sample - loss: 0.4502 - accuracy: 0.7960
Epoch 3/20
8000/8000 [==============================] - 1s 119us/sample - loss: 0.4376 - accuracy: 0.7960
Epoch 4/20
8000/8000 [==============================] - 1s 120us/sample - loss: 0.4308 - accuracy: 0.7960
Epoch 5/20
8000/8000 [==============================] - 1s 119us/sample - loss: 0.4255 - accuracy: 0.7965
Epoch 6/20
8000/8000 [==============================] - 1s 121us/sample - loss: 0.4195 - accuracy: 0.8021
Epoch 7/20
8000/8000 [==============================] - 1s 119us/sample - loss: 0.4127 - accuracy: 0.8180
Epoch 8/20
8000/8000 [==============================] - 1s 119us/sample - loss: 0.4050 - accuracy: 0.8244
Epoch 9/20
8000/8000 [==============================] - 1s 119us/sample - loss: 0.3959 - accuracy: 0.8291
Epoch 10/20
8000/8000 [=

# Model Evaluation and Prediction 

In [155]:
# evaluate the model
test_loss, test_acuracy =model.evaluate(x_test, y_test.to_numpy())

2000/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

In [157]:
print("Test accuracy: {}".format(100*test_acuracy))

Test accuracy: 85.85000038146973


Test accuracy and Train accuray is not much different, (no such a big gap) so it's good to go

In [158]:
y_pred = model.predict_classes(x_test)
y_pred

array([[0],
       [0],
       [0],
       ...,
       [0],
       [0],
       [0]], dtype=int32)

In [159]:
y_test

9394    0
898     1
2398    0
5906    0
2343    0
8225    1
5506    0
6451    0
2670    1
3497    1
1087    0
1819    0
2308    0
6084    0
3724    1
3184    1
6387    0
3728    0
2702    0
7883    0
2930    0
5988    0
4890    0
6718    0
5423    0
3213    0
3017    0
382     0
4237    0
4721    0
       ..
9236    0
1255    0
3445    0
5859    0
5450    0
5299    0
1712    0
7920    0
5194    0
1578    0
4943    0
3975    1
7134    0
5760    0
5276    0
8556    0
1624    0
4707    0
8387    0
5507    1
5581    0
7133    0
4041    0
4807    0
1774    0
1037    0
2899    0
9549    0
2740    0
6690    0
Name: Exited, Length: 2000, dtype: int64

Note second data has diffrent in y_pred and the y_test, it can be happend because it's accuray is around 85.85

In [161]:
# confusion matrics
from sklearn.metrics import confusion_matrix, accuracy_score
cm=confusion_matrix(y_test, y_pred)
print(cm)

[[1513   82]
 [ 201  204]]


In [163]:
acs=accuracy_score(y_test, y_pred)
print(100*acs)

85.85000000000001
